src/emulator/xlHeap.c

```diff
@@ -1,590 +1,645 @@
+    gnHeapTakeCount[iHeap] = 0;
 
-    for(i = 0; i < 11; i++) {
-        for(j = 0; j < 32; j++) {
-            gapHeapBlockCache[heap][i][j] = NULL;
+    for (nBlockSize = 0; nBlockSize < 11; nBlockSize++) {
+        for (nBlock = 0; nBlock < 32; nBlock++) {
+            gapHeapBlockCache[iHeap][nBlockSize][nBlock] = NULL;
         }
     }
 
-    tmp = gpHeapBlockFirst[heap];
-
-    while(NUM_WORDS(*tmp)) {
-        s32 tmp2 = NUM_WORDS(*tmp);
-
-        if(IS_FREE(*tmp)) {
-            xlHeapBlockCacheAdd(heap, tmp);
+    pBlock = gpHeapBlockFirst[iHeap];
+    while ((nBlockSize = BLOCK_SIZE(*pBlock)) != 0) {
+        if (BLOCK_IS_FREE(*pBlock)) {
+            xlHeapBlockCacheAdd(iHeap, pBlock);
         }
-
-        tmp += tmp2 + 1;
+        pBlock += nBlockSize + 1;
     }
 
-    return 1;
+    return true;
 }
 
-#ifdef NON_MATCHING
-s32 xlHeapCompact(s32 heap) {
-    s32 r8, r7, r6, r5, r4, r0;
-    u32 *cur = gpHeapBlockFirst[heap];
-    u32 *next;
-    u32 *prev = NULL;
-    s32 size;
-    u32 next_v;
-
-    while((size = *cur & 0xFFFFFF, size != 0)) {
-        next = cur + size + 1;
-        next_v = *next;
-
-        if(*cur & 0x1000000) {
-            if(r8 < size) {
-                r8 = size;
-            }
-
-            if(r7 < size) {
-                r7 = size;
-            }
+static inline bool xlHeapFindUpperBlock(s32 iHeap, s32 nSize, u32** ppBlock, s32* pnBlockSize) {
+    s32 nBlockSize;
+    u32 nBlock;
+    u32* pBlock;
+    u32* pBlockBest;
+    u32* pBlockNext;
+
+    pBlockBest = NULL;
+    pBlock = gpHeapBlockFirst[iHeap];
+
+    while ((u32)pBlock < (u32)gpHeapBlockLast) {
+        nBlock = *pBlock;
+        nBlockSize = BLOCK_SIZE(nBlock) & 0xFFFFFF;
+        if (CHKSUM_LO(nBlock) != CHKSUM_HI(nBlock)) {
+            return false;
+        }
+        if (BLOCK_IS_FREE(nBlock) && nBlockSize >= nSize) {
+            pBlockBest = pBlock;
+        }
+        pBlock += nBlockSize + 1;
+    }
 
-            if(r6 < size) {
-                r6 = size;
-            }
+    if (pBlockBest == NULL) {
+        return false;
+    }
 
-            if(r5 < size) {
-                r5 = size;
-            }
+    nBlockSize = BLOCK_SIZE(*pBlockBest);
+    xlHeapBlockCacheClear(iHeap, pBlockBest);
 
-            if(r4 < size) {
-                r4 = size;
-            }
+    if (nBlockSize > nSize + 0x20) {
+        pBlockNext = pBlockBest + ((nBlockSize - nSize) - 1);
+        *pBlockNext = MAKE_BLOCK(nSize, FLAG_FREE);
+        xlHeapBlockCacheAdd(iHeap, pBlockBest);
+        *ppBlock = pBlockNext;
+        *pnBlockSize = BLOCK_SIZE(*pBlockNext);
+    } else {
+        *ppBlock = pBlockBest;
+        *pnBlockSize = BLOCK_SIZE(*pBlockBest);
+    }
+    return true;
+}
 
-            if(r0 < size) {
-                r0 = size;
+// https://decomp.me/scratch/2K9NF
+#ifndef NON_MATCHING
+#pragma GLOBAL_ASM("asm/non_matchings/virtual_console/xlHeap/xlHeapCompact.s")
+#else
+bool xlHeapCompact(s32 iHeap) {
```

Comment:
matched: https://decomp.me/scratch/I6co7

changed `pBlockNext = pBlock + nBlockSize + 1` to `pBlockNext = pBlock + 1 + nBlockSize` lol

Reply:
I hate that it was this simple lol, thanks for matching this 🎉